source("./utils_codes/utils.R")
source("./utils_codes/downstream.R")
source("./utils_codes/normalize.R")
source("./utils_codes/hrmf_em.R")
source("./utils_codes/hrmf_init.R")
source("./utils_codes/image_tools.R")
# Spatial CNA Analysis Pipeline

#
# This function performs spatial CNA detection and visualization
# given a Seurat object and spatial transcriptomics data.
#
# @param sample_dir Directory of the sample data (contains `seurat_object.rds` and `gene_pos.rds`).
#   - `seurat_object.rds`: A Seurat object.
#       * Expression matrix stored in `obj@assays$Spatial@counts`.
#       * Contains clustering results from the Seurat standard workflow (in `obj$seurat_clusters`), 
#         or custom cell-type annotations that can be used to define normal reference cells.  
#   - `gene_pos.rds`: A data.frame containing genomic annotation of genes with the following columns:
#       * `symbol`: Gene symbol.
#       * `chr`: Chromosome name (e.g., "chr1").
#       * `start`: Start position of the gene (bp).
#       * `end`: End position of the gene (bp).
#
# @param BICseq_dir Path to the MBICseq executable.
#   - Default is "./code/MBICseq" (relative path inside the package/project).
#   - Users should provide the absolute path to the executable when running in their environment.
#   - MBICseq is used for segmentation in CNA calculation.
#
# @param image_dir Directory containing the image feature results generated by `get_spatial_feature.py`.  
#   The directory must include a subfolder `resnet50/` with the following files:  
#   - `features_pca.txt`: PCA-reduced image feature matrix.  
#   - `features_names.txt`: Feature names corresponding to the PCA features.  
#
# @param plot_dir Output directory for plots and results.
# @param bin_file Path to bin definition file (e.g., "bin_K=1e+06_hg38.RDS").
# @param normal_clusters IDs of clusters in `seurat_object.rds` (`obj$seurat_clusters`) 
#   that represent normal cells to be used as reference.
# @param state_ratio Copy number state ratios (default: c(0.5, 1, 1.5)).
# @param max_iter Maximum iterations for HMRF (default: 5).
# @param beta_fixed Beta parameter for HMRF (default: 5).
# @param update_gaussian Logical. Whether to update the mean and standard deviation of the Gaussian distribution.

# @param tumor_perc Numeric vector indicating the tumor fraction for spots. 
#   This can be estimated using SpaCNA's downstream module. Defaults to 1 for all spots.
#
# @return A list containing processed CNA matrices and intermediate results.

SpaCNA <- function(sample_dir,
                   image_dir,
                   plot_dir,
                   BICseq_dir='./utils_codes/MBICseq',
                   bin_file='./data/bin_K=1e+06_hg38.RDS',
                   data_type="10X",
                   normal_clusters,
                   dlm_dV=0.1, dlm_dW=0.001,
                   state_ratio = c(0.5,1,1.5),
                   max_iter = 5,
                   beta_fixed = 5,
                   tumor_perc=1,
                   update_gaussian=FALSE) {

    # load object
    obj <- readRDS(paste0(sample_dir,"seurat_object.rds"))

    ## inputs
    count_RNA <- as.matrix(obj@assays$Spatial@counts)
    count_RNA <- count_RNA[rowMeans(count_RNA)>0.05, ]
    message("Input count matrix dimension: ", paste(dim(count_RNA), collapse=" x "))

    bin <- readRDS(bin_file)
    cell_barcodes <- colnames(count_RNA)
    seurat_cluster <- as.numeric(obj$seurat_clusters)-1
    cluster <- as.character(seurat_cluster)
    spot_location <- get_spot_location(obj,type=data_type)
    normal_cells <- cell_barcodes[cluster %in% normal_clusters]

    # gene position
    gene_pos <- readRDS(paste0(sample_dir, "gene_pos.rds"))
    gene_pos <- gene_pos[gene_pos$symbol %in% rownames(count_RNA), ]
    gene_pos <- gene_pos[order(as.numeric(substr(gene_pos$chr,4,5)), gene_pos$start), ]
    gene_pos$n <- 1:nrow(gene_pos)

    # count gene number in bins
    gene_num <- rep(0, nrow(bin))
    for(i in 1:nrow(bin)){
        f <- gene_pos$chr==bin$chr[i] & gene_pos$start>=bin$start[i] & gene_pos$end<=bin$end[i]
        gene_num[i] <- sum(f)
    }

    ## gene smooth
    count_RNA <- count_RNA[gene_pos$symbol,]
    count_RNA_smooth <- chr_smooth(count_RNA, log=T, dlm_dV=dlm_dV, dlm_dW=dlm_dW)
    count_RNA_smooth <- edgeR::cpm(count_RNA_smooth)

    neigh_cell <- get_final_neigh(image_dir, spot_location, k=7, thre=0.2, dist_thre='unknown')
    count_RNA_spa <- spatial_smooth(count_RNA_smooth,
                                    neigh_cell,
                                    neigh_cell,
                                    alpha=c(0.5, 0.25, 0.25))
    count_RNA_norm <- get_copy_ratio(count_RNA_spa, normal_cells)

    count_norm <- generate_bin_count(count_RNA_norm, gene_pos, bin, func=mean)
    f <- gene_num >= 2
    count_norm <- count_norm[f,]
    if (!dir.exists(plot_dir)) {
        dir.create(plot_dir, recursive = TRUE)
    }
    saveRDS(count_norm, paste0(plot_dir, "count_norm.rds"))
    # plot CNA heatmap
    f <- sample(1:length(cluster), 1000)
    plot_heatmap(t(count_norm)[f,],
                cell_cluster=cluster[f],
                cell_annotation=cluster[f],
                output_dir=plot_dir,
                output_name="copy_ratio.png")

    ## hmrf para init
    parameter_init <- estimate_hmrf_parameter(count_RNA=count_RNA,
                                            normal_cells=normal_cells,
                                            gene_pos=gene_pos,
                                            bin=bin,
                                            gene_normalize=F,
                                            cnv_num=3,
                                            num_cells=200,
                                            tumor_perc=tumor_perc,
                                            common_dispersion=0.1,
                                            dropout_prob="au",
                                            dlm_dV=0.1,
                                            dlm_dW=0.002,
                                            alpha=c(0.5, 0.25, 0.25),
                                            gene_thre=2)
    saveRDS(parameter_init, paste0(plot_dir, "hrmf_para.rds"))

    ## cn state init
    mus <- parameter_init[[1]]
    sigmas <- parameter_init[[2]]
    mus_df <- data.frame(row.names=cell_barcodes,
                        p=rep(tumor_perc, length(cell_barcodes)),
                        mu1=mus[1], mu2=mus[2], mu3=mus[3])
    sigs_df <- data.frame(row.names=cell_barcodes,
                        p=rep(tumor_perc, length(cell_barcodes)),
                        mu1=sigmas[1], mu2=sigmas[2], mu3=sigmas[3])

    cnv_state <- get_cnv_init_perce(count_norm, mus_df, sigs_df)
    cnv_state_init <- state2ratio(cnv_state, state_ratio)
    f <- sample(1:length(cluster), 1000)
    plot_heatmap(t(cnv_state_init)[f,],
                cell_annotation=cluster[f],
                output_dir=plot_dir,
                output_name="cns_init.png")

    ## segment
    baseline <- rowMeans(count_norm[,normal_cells])

    bk_bic <- calculate_CNV(t(count_norm),
                        baseline,
                        lambda=0.005,
                        sample_dir=plot_dir,
                        BICseq_dir=BICseq_dir)
    saveRDS(bk_bic, paste0(plot_dir, "bk_bic.rds"))

    ## hmrf final
    cns <- get_state_icm_seg_perc(count_norm,
                                bk_bic,
                                neigh_list=neigh_cell,
                                mus_df=mus_df,
                                sigmas_df=sigs_df,
                                max_iter=max_iter,
                                beta_fixed=beta_fixed,
                                update_gaussian=update_gaussian)
    cns_ <- cns
    cns <- state2ratio(cns_, state_ratio)
    saveRDS(cns, paste0(plot_dir, "cns.rds"))

    cns <- denoise(cns, thre=0.98)
    plot_heatmap(t(cns),
                cell_cluster=cluster,
                cell_annotation=cluster,
                output_dir=plot_dir,
                output_name="cns.png")

    message("âœ… SpaCNA finished! Results saved in: ", plot_dir)

    return(list(
        count_norm = count_norm,
        parameter_init = parameter_init,
        cnv_state_init = cnv_state_init,
        bk_bic = bk_bic,
        cns = cns
    ))
}


# sample_dir <-  ""
# image_dir<- ""
# plot_dir <- ""

# cna_list<-SpaCNA(sample_dir,image_dir,plot_dir,normal_clusters = c(0,1,3,8))
